RFCs/0009-Fn::For.md

```diff
@@ -0,0 +1,903 @@
+# New Intrinsic Function for looping: `Fn::Map`
+
+* **Original Author(s):**: @mingujo
+* **Tracking Issue**: [Tracking Issue: Fn::Map](https://github.com/aws-cloudformation/cfn-language-discussion/issues/9)
+
+# Summary
+We are proposing to introduce a new intrinsic function called `Fn::Map`. It will allow users to iterate over a given list to replicate a particular template fragment thereby avoiding repetitively writing a similar declaration for each object (e.g. Resource, Output, or Resource Properties).
+
+`Fn::Map` traverses a single given List of Strings from first to last, and over each iteration, it references each item under a given template snippet to be replicated.
+
+We additionally introduce `Fn::Merge` which can be used to merge the result of Fn::Map into its parent object.
+# Motivation
+
+In a CloudFormation template, a single Resource configures one infrastructure object. Therefore, a template can become verbose when you manually declare multiple similar resources. Some examples would include a pool of EC2 instances with the same configurations but different instance types, or S3 Bucket Notification configurations with different SNS Topics.
+
+# Declaration
+## Fn::Map
+### Json
+```json
+{
+  "Fn::Map": {
+    "index": "Index",
+    "value": "Value",
```

Comment:
Things I'm wondering:
- Would it be clearer to use `IndexParameter` and `ValueParameter` for the keys (to make it clear that you're configuring something you will `!Ref` later)
- Would it be clearer to use `Item` instead of `Value`? This might be related to the languages I've been exposed to, but I'd use Key-Value or Index-Item, not using `Value` prevents confusion with the `Key` parameter

Reply:
I think "key" and "value" if they are both used should be *outputs*, currently "fragment" is used for the output value. We've got four parameters here:
* input value from the input collection (currently "value")
* index of the input value (currently "index")
* output value (currently "fragment")
* optional output key (currently "key")
Maybe making them more explicit makes sense? `InputIndex`, `InputValue`, `OutputValue`, `OutputKey`? It would leave open the possibility that in the future the collection could be an object, which would would then involve an `InputKey` parameter as well.